### ISSUE TITLE
[Withdrawn] SDL 0273 Revisions - WebEngine Projection mode

### ISSUE THREAD

**Issue by theresalech** [smartdevicelink/sdl_evolution#1064]
Hello SDL community,

The review of "Revise 0273-webengine-projection-mode" begins now and runs through August 11, 2020.

This will be a review of proposed revisions to a previously accepted but not yet implemented proposal, [SDL 0273](https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0273-webengine-projection-mode.md).  

The pull request outlining the revisions under review is available here:

https://github.com/smartdevicelink/sdl_evolution/pull/1063

Reviews are an important part of the SDL evolution process. All reviews should be sent to the associated Github issue at:

https://github.com/smartdevicelink/sdl_evolution/issues/1064

What goes into a review?

The goal of the review process is to improve the proposal under review through constructive criticism and, eventually, determine the direction of SDL. When writing your review, here are some questions you might want to answer in your review:

* Is the problem being addressed significant enough to warrant a change to SDL?
* Does this proposal fit well with the feel and direction of SDL?
* If you have used competitors with a similar feature, how do you feel that this proposal compares to those?
* How much effort did you put into your review? A glance, a quick reading, or an in-depth study?
Please state explicitly whether you believe that the proposal should be accepted into SDL.

More information about the SDL evolution process is available at

https://github.com/smartdevicelink/sdl_evolution/blob/master/process.md

Thank you,
Theresa Lech

Program Manager - Livio
theresa@livio.io

**Comment by kshala-ford** [comment]
I want to note that the concerning ambiguity may be due to miscommunication. In the original proposal we clearly describe the downsides or the impact to existing code due to the requirement:

> ### Potential downsides
> ....
> 2. The proposal states: If a WebEngine application attempts to register with this HMI type but the local policy table doesn’t allow, Core should not allow the app to register. This can be seen as a downside as it could break with apps being used for the first time. This feature would require the IVI to implement policy table updates through the vehicle modem instead of using a mobile application. It also requires the HMI to use SetAppProperties upon app installation to add the app policy ID to the policy table in order to trigger the policy table update through the mode.
> ...
> ### Impact to existing code
> ... Core would need to specially treat apps with this HMI type as they are not allowed to register unless permission was granted. ...

Now, an empty `AppHMIType` field is per our definition not allowing the use of `WEB_VIEW` because obviously `WEB_VIEW` is not part of the array. We decided independently of any current implementation and think the accepted solution provides the highest security against apps tampering with possible misconfiguration or states like pre-policy update.

Regarding proposal contents I am unsure if this revision is sufficient looking at the sections Potential downsides or Impact to existing code.

Regarding the proposed change I have following concerns:

1. human error allowing an app to use web view unintentionally because you can easily forget the HMI types in your SDL server
2. the policy database status when apps register which are unknown before would allow apps to operate with WEB_VIEW until PTU is complete
3. This is just to be clear and confirm with everyone that this revision does not have impact to it: This HMI type should only be allowed for WebEngine apps

If an app registers which is completely unknown, the AppHMIType is empty hence the app can successfully register with WEB_VIEW. In the meantime these apps would remain operational and cause a web view to be shown.

The proposal would need additional revisions to add a strategy to unregister these apps again. However, this is not sufficient because these apps should not be operational in the first place. I have thought about a strategy like "Waiting on Device" but I wanted to provide a strategy which is not too difficult to implement and provides a good user experience.

The original authors of the proposal are looking into it if this is revision is acceptable. Any feedback from the PM to the above items is greatly appreciated.





15:39

**Comment by JackLivio** [comment]
@kshala-ford Thank you for the review. Your comment does a good job of pointing out some complexities I may have missed. Let's examine closer.

1. > human error allowing an app to use web view unintentionally because you can easily forget the HMI types in your SDL server

I do not disagree with that, but the same could be said for forgetting to include the WEB_VIEW appHMIType. In general I think it is important for the SDL server to send the AppHMIType in all cases, especially now with the inclusion of WEB_VIEW.

2. > the policy database status when apps register which are unknown before would allow apps to operate with WEB_VIEW until PTU is complete

After writing out the following explanation, I can see that my revision is missing quite a few important details, so I apologize for that.

Anyways here is what I envisioned for the RAI request logic.

```
App registers with WEB_VIEW AppHMIType:

- If core does NOT have any app policy entry in the local PT for the web view app, then disallow the RAI.
- If core DOES have an app policy entry in the local PT for the web view app but is labeled as `predataConsent` or `default`, then disallow the RAI.
- If core DOES have a policy table entry for that app WITHOUT AppHMIType entry, allow RAI. (all types allowed)
- If core DOES have a policy table entry for that app WITH an AppHMIType entry, check that WEB_VIEW type is present. Allow RAI if WEB_VIEW is present. Disallow RAI if WEB_VIEW is NOT present.
```

The idea here is that Core would disallow RAI if it does not have a policy entry from the SDL serer, or if WEB_VIEW was not included in the AppHMIType array (if the AppHMIType array existed).

SDL Core sets labels on policy entries to track if an App has policies that were created from the `default` or `PreDataConsent` policy entries. When an app first registers it is either assigned preDataConsent or default policies (depends on the policy mode / external consent status). The default or preDataConsent label is cleared from the app's policy entry when a policy table update occurs and the update included specific policies for that app.

When an app policy is created from the HMI SetAppProperties rpc, the is given default policy permissions, but I need to test to see if it also receives the default policy label.


[The label for an app policy can be checked via `get_string()`](https://github.com/smartdevicelink/sdl_core/blob/master/src/components/rpc_base/include/rpc_base/rpc_base_inl.h#L688). This is a way to see if an app is using the default or predataconsent policies. BTW the corresponding setter method is called `set_to_string()` if you wanted to look at how those methods are used.

3. > This is just to be clear and confirm with everyone that this revision does not have impact to it: This HMI type should only be allowed for WebEngine apps

That is my understanding. The revision is only to address the specifics of how policies enforce the WEB_VIEW appHMIType usage.

> If an app registers which is completely unknown, the AppHMIType is empty hence the app can successfully register with WEB_VIEW. In the meantime these apps would remain operational and cause a web view to be shown.

I think this is addressed in my response to 2. If an app has default or preDataConsent policies, then do not allow it to register with `AppHMIType::WEB_VIEW`.

**Comment by kshala-ford** [comment]
@JackLivio 

>> human error allowing an app to use web view unintentionally because you can easily forget the HMI types in your SDL server
> I do not disagree with that, but the same could be said for forgetting to include the WEB_VIEW appHMIType. In general I think it is important for the SDL server to send the AppHMIType in all cases, especially now with the inclusion of WEB_VIEW.

The question is which option is the more critical one. My position is: permission to sensitive features should be given explicitly.

Looking at your four bullet points I believe the revision is only related to the third bullet point:

> - If core DOES have a policy table entry for that app WITHOUT AppHMIType entry, allow RAI. (all types allowed)

The original proposal would not allow app registration in this case. This would require a specific implementation whereby your revision is suggesting consistent behavior to other HMI types which also requires specific implementations. 

The check if `WEB_VIEW` is contained in the policy entry would be sufficient for the other bullet points as well because the hmi type is obviously not present in these scenarios.

There will be a one-off solution for the `WEB_VIEW` HMI type with or without the revision. The original proposal is more strict not allowing `WEB_VIEW` usage unless it's explicitly permitted. According to how many concerns were brought up during the review of the original proposal, I think the SDLC members should support the solution which is safer against mistakes and misuse.

**Comment by joeygrover** [comment]
I could be missing something but here is my thought process. Our main goal of the original policy table changes was to prevent apps that are not approved to use the `WEB_VIEW` from doing so.

I think that the technical limitations around the feature itself actually add a level of prevention and render the policy table changes a bit unnecessary. Examining a few things we know:

1\. We know that in order to use the `WEB_VIEW` an app must be a web engine type app. This means that they need at least some code installed onto the head unit. This would include the manifest and binaries of the app.

2\. In order for an app to be installed on the IVI system, the IVI will have to have some sort of manager to handle the downloading, installation, etc. This has been referred to as an "app store". This is currently out of the scope of SDL and it is up to the OEM to implement their own version of the app store even if it uses the sample one provided in the GENERIC_HMI project. But we will make an assumption that this app store obtains its app list from an OEM server that returns approved applications.

3\. The HMI is responsible for launching web engine applications and this flow is outside the scope of SDL. Most OEMs will use something like the previously mentioned app store to know how to launch the application. This will include likely starting either an entire web engine page or iframe.

So let's look at a few use cases:

- If an application registers with `WEB_VIEW`, the HMI will attempt to launch the application based upon the app store implementation as mentioned in point 3.  If the application is not listed in the app store the HMI should not launch the web engine and the app activation would fail. Therefore, no application outside web engine applications (Android, iOS, Java, Node.js) should be able to activate because of this simple check.
- Web engine apps that are not supposed to use the `WEB_VIEW` would get past this initial check. However, because of point 2 we know that in order for this app to be installed on the system it must have been included into the OEM's app store. Which means the app would have already been approved by that OEM and this corner case should not occur. 
- In the off chance a web engine app gets installed from the app store that somehow tries to access the `WEB_VIEW` the app can be removed by the OEM from the app store and have it uninstalled. And because this is the only real outlier that might cause the issue we are concerned about we have to ask is the ad-hoc change worth the effort of making the change.

**Comment by LuxoftAKutsan** [comment]
I would like to post the following arguments into https://github.com/smartdevicelink/sdl_evolution/issues/1064, but not sure if this would be OK.
I would not like to make the discussion even more protracted.

    I think this is addressed in my response to 2. If an app has a default or preDataConsent policies, then do not allow it to register with `AppHMIType::WEB_VIEW`.

@JackLivio
Such implementation is possible on the sdl_core side, but for now, we don't have similar use cases that use such trick.
Implementation of this behavior requires additional time, and there is a risk that SDL may contain some amount of bugs that will make this implementation even more complicated.
I would leave behavior as we already voted in the accepted version of the proposal.
Main arguments :

 - The behavior mentioned in the proposal multiple times and we all get agreed on it.
 - The proposal has high priority to be included in the 7.0 release, we would like to avoid the risk of excluding it from release with additional implementation.

@joeygrover 
Sure, Web engine projection apps should be approved by OEMs before it will bring it to the head unit.
But HMI type is the value that the app sends in the Register app interface, so it can be changed in application run time.
OEM probably will not have enough resources for deep source code review and testing to find all possible app behavior.
Also, the WebEngine projection apps may have access to the internet and may change its behavior depends on external commands. Therefore, I wouldn't make Web engine apps more trusted then mobile because they are executed on HU.
SDL should apply the same policy mechanism should prevent unwanted behavior of the app.
And SDL should not allow registering Web engine apps without explicit policy records with appropriate access.

**Comment by LuxoftAKutsan** [comment]
I would like to post the following arguments into https://github.com/smartdevicelink/sdl_evolution/issues/1064, but not sure if this would be OK.
I would not like to make the discussion even more protracted.

    I think this is addressed in my response to 2. If an app has a default or preDataConsent policies, then do not allow it to register with `AppHMIType::WEB_VIEW`.

@JackLivio
Such implementation is possible on the sdl_core side, but for now, we don't have similar use cases that use such a trick.
Implementation of this behavior requires additional time, and there is a risk that SDL may contain some amount of bugs that will make this implementation even more complicated.
I would leave behavior as we already voted in the accepted version of the proposal.
Main arguments :

 - The behavior mentioned in the proposal multiple times and we all get agreed on it.
 - The proposal has a high priority to be included in the 7.0 release, we would like to avoid the risk of excluding it from release with additional implementation.

@joeygrover 
Sure, Web engine projection apps should be approved by OEMs before it will bring it to the head unit.
But HMI type is the value that the app sends in the Register app interface, so it can be changed in application run time.
OEM probably will not have enough resources for deep source code review and testing to find all possible app behavior.
Also, the WebEngine projection apps may have access to the internet and may change its behavior depends on external commands. Therefore, I wouldn't make Web engine apps more trusted then mobile because they are executed on HU.
SDL should apply the same policy mechanism should prevent unwanted behavior of the app.
And SDL should not allow registering Web engine apps without explicit policy records with appropriate access.

**Comment by kshala-ford** [comment]
@joeygrover We want to take advantage of the SDL policy system which includes protecting apps from invalid registrations. You mention the manifest where `category` is a mandatory field and must be one of the elements in `AppHMIType`. Things are getting quite complicated as the manifest content needs to be aligned with the policy server and of course the local policy snapshot. In our AppStore we would expect SDL Core to prevent apps registering with any invalid HMI type especially with `WEB_VIEW`. Our AppStore would compare the content of `HMIApplication` with that manifest content to see if the web view is desired or not. With this revision we would need to find a way to de-register the application and add logic for notification and reporting so that we can identify this behavior (the SDL policy table does include reporting). We would be in a situation like with mobile navigation apps not knowing if the app is allowed to be a navigation apps or not. We tried to avoid this in the original proposal as it's complex and error prone. 

Again. I don't see the reason to avoid a single and easy one-off condition but add two conditions which additionally compare functional groups. The protecting logic is desired in SDL. Otherwise the biggest reason to use WebEngine apps with SDL needs to be revisited if the PM expects the AppStore to take over the job.

**Comment by JackLivio** [comment]
@LuxoftAKutsan @kshala-ford Ok I will withdraw the suggested changes outlined in the revision and in my previous comments. I don't believe this revision to be a big enough reason to risk the Webengine delivery for 7.0 since it doesn't seem we will come to an agreement on the change this week. 

Fixing inconsistencies in the policy table is in our evolution backlog so we may revisit this conversation in a future proposal. Such changes go beyond the scope of the webengine proposal.

**Comment by theresalech** [comment]
This proposal has been withdrawn per the author's request.
